### PR TITLE
Optionally allow learning incognito mode 1367

### LINKED ISSUE #1367: Add option to allow learning in Incognito/Private Browsing modes
My apologies if this is the way Privacy Badger is supposed to work, or this functionality is not possible to add, but I noticed that Privacy Badger does not add rules when I am in incognito mode.

These are the steps I performed to test this.

1. Deleted user data directory so Privacy Badger is in its default state of not blocking anything.
2. Visited nytimes.com and wsj.com in incognito mode - no rule changes were made.
3. Visited the same sites in regular mode - Privacy Badger added rules to block several domains.

Browser: Google Chrome 58.0.3029.96
OS: Ubuntu 16.04

### PR DESCRIPTION
**Pull request by mijiturka**
Resolves #1367.

If the user chooses to, treat all windows as if they are regular ones and keep records on them for heuristics.
This is disabled by default because it results in the storage of user history during a private session.

* Renamed tabIsIncognito to dontLearnInTab to better reflect the new value it returns (false for incognito windows if the user has allowed learning in them).
* Updated Bulgarian translation.

### CODE REVIEW

**Comment by mijiturka** [comment]
The build is failing because of the lint which disallows undefined usages of the "badger" instance - but incognito.js is required by other files that have the instance. How should I solve this? I saw other files like storage.js and heuristicblocking.js using a badger instance without passing it around (or as far as I understand, including it as part of their requirements).

**Comment by ghostwords** [comment]
`badger` is a global var [we define in `background.js`](https://github.com/EFForg/privacybadger/blob/43143a185c9074d8223fe6537a314465ccccba88/src/js/background.js#L874). Since `background.js` drives all the [other background-related files](https://github.com/EFForg/privacybadger/blob/43143a185c9074d8223fe6537a314465ccccba88/src/manifest.json#L25-L42), by the time we get to executing the code in those files, `badger` should be defined. However, as it's an implicitly available global var, we have to tell ESLint about it once per file ([for example](https://github.com/EFForg/privacybadger/blob/43143a185c9074d8223fe6537a314465ccccba88/src/js/storage.js#L18)), or ESLint will rightly complain.

I realize this isn't intuitive ... we could probably tweak something somewhere to make using `badger` more obvious (without switching to a real loader (#1612), which doesn't seem worth it at this time). Maybe register `badger` with [require.scopes](https://github.com/EFForg/privacybadger/blob/43143a185c9074d8223fe6537a314465ccccba88/src/js/bootstrap.js#L37) and then replace all the ESLint hints with explicit requiring (`let badger = require("badger");`)?

File: src/skin/options.html [hunk 1]
```diff
@@ -161,9 +161,17 @@ <h1><span class="i18n_options_title"></span></h1>
...
+          <input type="checkbox" id="learn_in_incognito_checkbox">
+          <span class="i18n_options_incognito_setting"></span>
+        </label>
+      </div>
       <div id="settingsSuffix">
-        <br>
+        <br />
         <span class="i18n_options_webrtc_warning"></span>
+        <br /><br />
+        <span class="i18n_options_incognito_warning"></span>
```
Comment (ghostwords) [inline src/skin/options.html]: Does this do the right thing in Firefox 52 (ESR) where WebRTC isn't enabled and so `"i18n_options_webrtc_warning"` above will be hidden? Looks like [all of `#settingsSuffix` gets hidden](https://github.com/EFForg/privacybadger/blob/43143a185c9074d8223fe6537a314465ccccba88/src/js/options.js#L121), so we should fix this.

File: src/_locales/en_US/messages.json [hunk 2]
```diff
@@ -199,9 +199,17 @@
         "message": "You're now protected by Privacy Badger. To learn how Privacy Badger works, click below for a quick tutorial.",
         "description": ""
     },
+    "options_incognito_setting": {
+        "message": "Learn in Private/Incognito windows **",
+        "description": "Checkbox label on the general settings page"
+    },
+    "options_incognito_warning": {
+        "message": "** Learning which domains are trackers in Private/Incognito windows will leave traces of your private browsing history on your computer. By default, Privacy Badger will still block trackers it already knows about in incognito windows - but you can enable this option to allow it to learn about new trackers and store information on them too.",
```
Comment (ghostwords) [inline src/_locales/en_US/messages.json]: How about the following?
>Enabling learning in Private/Incognito windows may leave traces of your private browsing history on your computer. By default, Privacy Badger will block trackers it already knows about in Private/Incognito windows, but it won't learn about new trackers. You might want to enable this option if most of your browsing happens in Private/Incognito windows.

File: src/js/incognito.js [hunk 3]
```diff
@@ -23,14 +23,18 @@ function startListeners() {
   chrome.tabs.onRemoved.addListener(onRemovedListener);
 }
 
-function tabIsIncognito(tabId) {
+function dontLearnInTab(tabId) {
```
[the remark this answers is not in the document]
Comment (mijiturka) [reply src/js/incognito.js]: You are absolutely right! Fixed now.

File: src/js/background.js [hunk 4]
```diff
@@ -682,6 +683,13 @@ Badger.prototype = {
     return self.getSettings().getItem("webRTCIPProtection");
   },
 
+  /**
+   * Check if learning about trackers in incognito windows is enabled
+   */
+  isLearnInIncognitoEnabled: function() {
```
[the remark this answers is not in the document]
Comment (mijiturka) [reply src/js/background.js]: I made it this way to keep it consistent with all the other options in the General tab. The only one that does something more involved then getting the settings item is isWebRTCIPProtectionEnabled. Do you think it's worth modifying all the rest to use getItem() instead?

File: src/skin/options.html (hunk 1 above)
Comment (mijiturka) [reply src/skin/options.html]: I missed this completely! Fixed now.

File: src/_locales/en_US/messages.json (hunk 2 above)
Comment (mijiturka) [reply src/_locales/en_US/messages.json]: I used this but replaced "most of your browsing" with "a lot of browsing" at the end. What do you think?
I don't want this option to sound scary or make people wonder if enough of their browsing happens in Incognito mode. It's really a question of whether you're using a shared computer or not.

**Comment by mijiturka** [comment]
I see about the badger global variable now! I will play around with it but I don't really understand it enough yet so a bit scared to touch it.

File: src/js/background.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (ghostwords) [reply src/js/background.js]: I feel fine about being inconsistent and fixing this sort of thing as we go. You could do it now if you like.

**Comment by ghostwords** [comment]
I definitely think this is a nit we can improve. `const badger = require("badger");` is more clear than `/* globals badger:false */`, plus fewer globals are always a good think. Do feel free to open a separate pull request. I can help if you have any questions.

File: src/js/options.js [hunk 5]
```diff
@@ -118,7 +118,7 @@ function loadOptions() {
   } else {
     // Hide WebRTC-related settings for non-supporting browsers
     $("#webRTCToggle").css({"visibility": "hidden", "height": 0});
-    $("#settingsSuffix").css({"visibility": "hidden", "height": 0});
+    $("#webrtc_warning").css({"visibility": "hidden", "height": 0});
```
Comment (ghostwords) [inline src/js/options.js]: This is pretty minor, but it seems like we get extra space above `#incognito_warning` when `#webrtc_warning` is hidden. I think it's from all the `<br>`s. The fix might be to remove the `<br>`s and set appropriate padding between `<span>`s inside `#settingsSuffix`.

**Comment by ghostwords** [comment]
This is related to #1899.

**Comment by ghostwords** [comment]
#1929 is a good followup issue (discovered while reviewing this PR).

**Comment by ghostwords** [comment]
Let me know if you'll have some time to look this over once more before I merge.

**Comment by ghostwords** [comment]
Added new strings to Transifex, downloaded and committed updated source and Bulgarian locales, updated all other locales with new strings.

**Pull request merged**